src/msix/pack/XmlWriter.cpp

```diff
@@ -0,0 +1,83 @@
+//
+//  Copyright (C) 2019 Microsoft.  All rights reserved.
+//  See LICENSE file in the project root for full license information.
+//
+#include "XmlWriter.hpp"
+#include "StringStream.hpp"
+#include "ComHelper.hpp"
+#include "Exceptions.hpp"
+#include "MsixErrors.hpp"
+
+#include <stack>
+#include <string>
+
+namespace MSIX {
+
+        const static char* xmlStart = "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"true\"?><";
+
+        // Adds xml header declaration plus the name of the root element
+        XmlWriter::XmlWriter(const std::string& root)
+        {
+            m_elements.emplace(root);
+            m_stream = ComPtr<IStream>::Make<StringStream>();
+            std::string start = xmlStart + root;
+            ULONG copy;
+            ThrowHrIfFailed(m_stream->Write(static_cast<const void*>(start.data()), static_cast<ULONG>(start.size()), &copy));
+            m_state = State::OpenElement;
+        }
+
+        void XmlWriter::StartElement(const std::string& name)
+        {
+            ThrowErrorIf(Error::XmlError, m_state == State::Finish, "Invalid call, xml already closed");
+            m_elements.emplace(name);
+            std::string element;
+            // If the state is open, then we are adding a child element to the previous one. We need to close that element's
+            // tag and add the new one. If the state is closed, there is no need to close the tag as it had already been closed.
+            if (m_state == State::OpenElement)
+            {
+                // close parent element
+                element = "><" + name;
+            }
+            else // State::ClosedElement
+            {
+                element = "<" + name;
+            }

```

Comment:
nit: I find it easier to follow if any common code is not inside conditionals.  Thus:

if (open)
  >

< + name